dart/constraint/ConstraintSolver.cpp

```diff
@@ -523,6 +551,19 @@ void ConstraintSolver::updateConstraints()
   // Add the new contact constraints to dynamic constraint list
   for (const auto& contactConstraint : mContactConstraints)
   {
+    // update the slip compliances of the contact constraints based on the
+    // number of contacts between the collision objects.
+    auto& contact = contactConstraint->getContact();
+    std::size_t numContacts = 1;
+    auto it = contactPairMap.find(
+        std::make_pair(contact.collisionObject1, contact.collisionObject2));
+    if (it != contactPairMap.end())
+      numContacts = it->second;
+
+    contactConstraint->setPrimarySlipCompliance(
+        contactConstraint->getPrimarySlipCompliance() * numContacts);
+    contactConstraint->setSecondarySlipCompliance(
+        contactConstraint->getSecondarySlipCompliance() * numContacts);
```

Comment:
Could you let me know why we need to multiply the slip compliance by the number of contacts?

If I understand correctly, the slip effect will be applied per contact anyway when there were multiple contacts for the same object pair. Suppose we have two contacts for the same object pair and multiply the slip compliance by two, then the _doubled_ slip effect will happen at the two contact points.

Reply:
A similar change was made in https://github.com/osrf/gazebo/commit/402b9d361ccae9e250392526eebd5cc987a8b814 (see also [bitbucket PR 2965](https://osrf-migration.github.io/gazebo-gh-pages/#!/osrf/gazebo/pull-requests/2965/page/1)). The analogy between slip and linear damping is useful here as well. Each contact point with slip acts like an extra damper acting in parallel, which increases the total slip acting on the body. This multiplication aims to make the total slip invariant of the number of contact points.